### ISSUE TITLE
Custom Middleware Breaks UseStatusCodePagesWithReExecute

### ISSUE THREAD

**Issue by nfplee** [dotnet/aspnetcore#25532]
I have the following middleware which allows me to change the content after the pipeline has executed:

```
public async Task InvokeAsync(HttpContext httpContext) {
    // Store the "pre-modified" response stream.
    var originBody = httpContext.Response.Body;

    // Set the response body to our stream so we can read after the chain of middleware has been called.
    using var body = new MemoryStream();
    httpContext.Response.Body = body;

    try {
        await _next(httpContext);
    } finally {
        httpContext.Response.Body = originBody;
    }

    var content = Encoding.UTF8.GetString(body.ToArray());

    // Replace content here

    // Write the content to the response stream.
    await httpContext.Response.WriteAsync(content);
}
```

I need this placed after `UseRequestLocalization` which itself must be placed after `UseRouting`. However `UseStatusCodePagesWithReExecute` must go before UseRouting. Therefore I end up with the following sequence:

1. UseStatusCodePagesWithReExecute
2. UseRouting
3. UseRequestLocalization
4. My middleware

The problem now is that my status code error page won't work. If I move my middleware to execute before UseStatusCodePagesWithReExecute then the error page works but the wrong culture is set as I need it placed after UseRequestLocalization.

I'd appreciate it if somone could help as I'm going around in circles trying to fix this.

**Comment by nfplee** [comment]
I've attached the following application to show you the problem.

[WebApplication1.zip](https://github.com/dotnet/aspnetcore/files/5162047/WebApplication1.zip)

If you open the Startup.cs file you'll see the default middleware order. Here's the different scenarios:

1. When you access a page which does not exist (e.g. /foo) the page not found view is not shown.
2. To fix this move UseMiddleware<MyMiddleware>() above UseStatusCodePagesWithReExecute(...). Now the page not found page is correctly shown. However if you go to /fr and add a breakpoint in the middleware you'll see the culture will give the default culture.
3. Tried to add UseRequestLocalization() above UseMiddleware<MyMiddleware>() but this fails as the culture is determined by the route and therefore it must come after UseRouting.
4. Next I tried adding UseRouting() above UseRequestLocalization() but now the page not found page is broken again. This is because UseStatusCodePagesWithReExecute must be added before UseRouting so the request can be rerouted to the correct page (as described in the documentation).
